### PR TITLE
QuickBuilder refactor

### PR DESCRIPTION
**Pull request by cristianoc72**
Refactor`QuickBuilder` to build classes on virtual filesystem.
This commit introduces the following differences:

-  The built classes are grouped in a single file and written on virtual filesystem, by default. Optionally, it's possible to write them on physical filesystem, for debugging purpose. When we use the physical filesystem, the classes are written in separated files, to make the debug easier. E.g.:
```php
<?php
use Propel\Generator\Util\QuickBuilder;

$schema = '<database>.............</database>';

//---The following builds on virtual filesystem
$builder = new QuickBuilder();
$builder->build($schema);
     // or alternatively
QuickBuilder::buildSchema($schema);


//---The following builds on physical filesystem
$builder = new QuickBuilder();
$builder->setVfs(false);
$builder->build($schema);
     // or alternatively
QuickBuilder::buildSchema($schema, null, null, null, null, false);
```
-  I moved `VfsTrait` from __Tests__  to __Propel\Generator\Util__ namespace, because otherwise PhpStan can't autoload it, rising an error (the more I use PhpStan the more I hate it :smile: ).
-   The refactored code and relative tests are strict typed.

### CODE REVIEW

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 1]
```diff
@@ -235,9 +274,9 @@ public function build($dsn = null, $user = null, $pass = null, $adapter = null,
     /**
      * @return \Propel\Generator\Model\Database|null
      */
-    public function getDatabase()
+    public function getDatabase(): ?Database
     {
-        if ($this->database === null) {
+        if (!isset($this->database)) {
```
Comment (stereomon) [inline src/Propel/Generator/Util/QuickBuilder.php:279]: What's the reason behind this change? !isset is slower than === null

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 2]
```diff
@@ -174,26 +187,52 @@ public function setConfig(GeneratorConfigInterface $config)
      */
     public function getConfig()
     {
-        if ($this->config === null) {
+        if (!isset($this->config)) {
```
Comment (stereomon) [inline src/Propel/Generator/Util/QuickBuilder.php:190]: What's the reason behind this change? `!isset` is slower than `=== null`

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 3]
```diff
@@ -144,9 +157,9 @@ public function setPlatform($platform)
      *
      * @return \Propel\Generator\Platform\PlatformInterface
      */
-    public function getPlatform()
+    public function getPlatform(): PlatformInterface
     {
-        if ($this->platform === null) {
+        if (!isset($this->platform)) {
```
Comment (stereomon) [inline src/Propel/Generator/Util/QuickBuilder.php:162]: What's the reason behind this change? `!isset` is slower than `=== null`

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 4]
```diff
@@ -114,17 +127,17 @@ public function getSchemaName()
...
 
     /**
-     * @return \Propel\Generator\Reverse\SchemaParserInterface
+     * @return \Propel\Generator\Reverse\SchemaParserInterface|null
      */
-    public function getParser()
+    public function getParser(): ?SchemaParserInterface
     {
-        return $this->parser;
+        return $this->parser ?? null;
```
Comment (stereomon) [inline src/Propel/Generator/Util/QuickBuilder.php:140]: Isn't `$this->parser` `null` by default?

Comment (dereuromark) [reply src/Propel/Generator/Util/QuickBuilder.php:140]: The docblock in property needs to be fixed then to append `|null`, as well

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 3 above)
Comment (dereuromark) [reply src/Propel/Generator/Util/QuickBuilder.php:162]: I would also like to avoid seeing any unrelated change here that is not necessary as it only blocks moving forward.

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 5]
```diff
@@ -343,11 +382,11 @@ public function getSQL()
     }
 
     /**
-     * @param string[]|null $classTargets
+     * @param array $classTargets
      *
      * @return string
      */
-    public function getBuildName($classTargets = null)
+    public function getBuildName(array $classTargets = []): string
```
Comment (dereuromark) [inline src/Propel/Generator/Util/QuickBuilder.php:389]: This accepted null so far.
I am not against making this a bit more strict - but we need to be careful about either the BC approach or to clearly state this in the release notes for the next release coming out.

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 6]
```diff
@@ -357,7 +396,7 @@ public function getBuildName($classTargets = null)
             $tables[] = $table->getName();
         }
         $name = implode('_', $tables);
-        if (!$classTargets || count($classTargets) === 5) {
+        if ($classTargets === [] || count($classTargets) === 5) {
```
Comment (dereuromark) [inline src/Propel/Generator/Util/QuickBuilder.php:399]: This can stay the way it was.

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 7]
```diff
@@ -606,8 +597,64 @@ public function isIdentifierQuotingEnabled()
...
-    public function setIdentifierQuoting($identifierQuoting)
+    public function setIdentifierQuoting(bool $identifierQuoting): void
     {
         $this->identifierQuoting = $identifierQuoting;
     }
+
+    /**
+     * Create separate classes to write to physical filesystem.
+     *
+     * @param array $classes
```
Comment (dereuromark) [inline src/Propel/Generator/Util/QuickBuilder.php:608]: You can be more concrete here, e.g. string[]
Same for other array usually

File: src/Propel/Generator/Util/QuickBuilder.php [hunk 8]
```diff
@@ -606,8 +597,64 @@ public function isIdentifierQuotingEnabled()
...
+            }
+        }
+
+        return $includes;
+    }
+
+    /**
+     * Create an all-classes file to write to virtual filesystem.
+     *
+     * @param array $classes
```
Comment (dereuromark) [inline src/Propel/Generator/Util/QuickBuilder.php:636]: Same

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 4 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php:140]: Nope. Starting from PHP 7.4, when you use strict types, the properties aren't initialize to null anymore. If `$this->parser` is not set, it's `uninitialized` and can't be accessed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 3 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php:162]: Starting from PHP 7.4, when you use strict types, the properties aren't initialized to null anymore, so  testing `=== null` rises a `Fatal error: Uncaught Error: Typed property QuickBuilder::$platform must not be accessed before initialization`.

Anyway, I can explicitly initialize the variables to `null`, if you think it's better.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 2 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php:190]: idem

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 1 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 2 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 3 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: I adjusted the code to change it as little as possible.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 5 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 7 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 8 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 4 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: I adjusted the code to change it as little as possible.

File: src/Propel/Generator/Util/QuickBuilder.php (hunk 6 above)
Comment (cristianoc72) [reply src/Propel/Generator/Util/QuickBuilder.php]: Fixed.

**Review by stereomon — APPROVED** [review]
